bakery/src/scripts/copy-resources-s3.py

```diff
@@ -0,0 +1,256 @@
+    session = boto3.session.Session()
+    s3_client = session.client(
+        's3', aws_access_key_id=aws_key, aws_secret_access_key=aws_secret)
+    s3_client.upload_file(
+        Filename=filename,
+        Bucket=bucket,
+        Key=key,
+        ExtraArgs={
+            "ContentType": content_type
+        }
+    )
+    return key
+
+
+def upload(in_dir, bucket, bucket_folder):
+    """ upload resource and resource json to S3 """
+    aws_key = os.getenv('AWS_ACCESS_KEY_ID')
+    aws_secret = os.getenv('AWS_SECRET_ACCESS_KEY')
+
+    disable_deep_folder_check = is_s3_folder_empty(
+        aws_key=aws_key,
+        aws_secret=aws_secret,
+        bucket=bucket,
+        key=bucket_folder)
+
+    # environment variables mandatory for ThreadPoolExecutor
+    if aws_key is None or aws_secret is None:
+        raise EnvironmentError(
+            'Failed because AWS_ACCESS_KEY_ID and/or AWS_SECRET_ACCESS_KEY environment variable missing.')
+
+    sha1_filepattern = '?' * 40  # sha1 hexdigest as 40 characters
+
+    # get all resources on file system
+    all_resources = []
+    for sha1filename in Path(in_dir).glob(sha1_filepattern):
+        resource = {}
+        resource['bucket'] = bucket
+        resource['input_file'] = str(sha1filename)
+        resource['input_metadata_file'] = resource['input_file'] + '.json'
+        resource['output_s3'] = slash_join(
+            bucket_folder, os.path.basename(resource['input_file']))
+        resource['output_s3_metadata'] = slash_join(bucket_folder,
+                                                    os.path.basename(resource['input_file']) + '-unused.json')
+        all_resources.append(resource)
+
+    # check which resources are missing (with ThreadPoolExecutor)
+    start = timer()
+    upload_resources = []
+    check_futures = []
+    with ThreadPoolExecutorStackTraced(max_workers=MAX_THREAD_CHECK_S3) as executor:
+        print('Checking which files to upload ', end='', flush=True)
+        for resource in all_resources:
+            check_futures.append(
+                executor.submit(
+                    check_s3_existence,
+                    aws_key=aws_key,
+                    aws_secret=aws_secret,
+                    bucket=bucket,
+                    resource=resource,
+                    disable_check=disable_deep_folder_check)
+            )
+        for future in concurrent.futures.as_completed(check_futures):
+            try:
+                resource = future.result()
+                if resource is not None:
+                    upload_resources.append(resource)
+                    if not disable_deep_folder_check:
+                        print('.', end='', flush=True)
+                else:
+                    if not disable_deep_folder_check:
+                        print('x', end='', flush=True)
+                check_futures.remove(future)
```

Comment:
Good question. It seemed to be a best practice on many examples on the internet found.
After investigation it seemed to be because of memory consumption. After a future's result is processed it's not needed anymore in the memory and keeps the memory consumption low:
https://stackoverflow.com/questions/34770169/using-concurrent-futures-without-running-out-of-ram

Actually I will add a line after this and `del future` itself manually too like on this examples https://www.programcreek.com/python/example/86022/concurrent.futures.as_completed